chainer/functions/connection/convolution_nd.py

```diff
@@ -0,0 +1,354 @@
+        n, c = x.shape[:2]      # (n, c_I, d_1, d_2, ..., d_N)
+        dims = x.shape[2:]
+        stride = self.stride
+        pad = self.pad
+        ndim = self.ndim
+        colon = slice(None)
+
+        # Compute output image's dimensions.
+        outs = tuple(
+            [conv.get_conv_outsize(d, k, s, p, cover_all=self.cover_all)
+             for (d, k, s, p) in zip(dims, ksize, stride, pad)])
+
+        # Make empty array for result.
+        y_shape = (n, out_c) + outs  # (n, c_O, out_1, out_2, ..., out_N)
+        y = cuda.empty(y_shape, dtype=x.dtype)
+        # Implementation using cuDNN.
+        if (not self.cover_all and cuda.cudnn_enabled and self.use_cudnn and
+            ndim > 1 and convolution_2d._check_cudnn_acceptable_type(
+                x.dtype, W.dtype)):
+            x = cuda.cupy.ascontiguousarray(x)
+            W = cuda.cupy.ascontiguousarray(W)
+            if b is not None:
+                b = cuda.cupy.ascontiguousarray(b)
+
+            handle = cudnn.get_handle()
+            x_desc = cudnn.create_tensor_descriptor(x)
+            y_desc = cudnn.create_tensor_descriptor(y)
+
+            self.filter_desc = cudnn.create_filter_descriptor(W)
+            self.conv_desc = cudnn.create_convolution_descriptor(
+                pad, stride)
+            if b is not None:
+                b_index = (None, colon) + (None,) * ndim
+                self.bias_desc = cudnn.create_tensor_descriptor(b[b_index])
+
+            workspace_size = cuda.get_max_workspace_size()
+            workspace = cuda.empty((workspace_size,), dtype='b')
+            algo = libcudnn.getConvolutionForwardAlgorithm(
+                handle, x_desc.value, self.filter_desc.value,
+                self.conv_desc.value, y_desc.value, _fwd_pref,
+                workspace_size)
+
+            oz_dtype = 'd' if x.dtype == 'd' else 'f'
+            one = numpy.array(1, dtype=oz_dtype).ctypes
+            zero = numpy.array(0, dtype=oz_dtype).ctypes
+            libcudnn.convolutionForward(
+                handle, one.data, x_desc.value, x.data.ptr,
+                self.filter_desc.value, W.data.ptr, self.conv_desc.value,
+                algo, workspace.data.ptr, workspace_size, zero.data,
+                y_desc.value, y.data.ptr)
+
+            # TODO(takagi) Support unshared bias
+            if b is not None:
+                cudnn.add_tensor(
+                    handle, one.data, self.bias_desc.value, b.data.ptr,
+                    one.data, y_desc.value, y.data.ptr)
+        # Implementation using im2col.
+        else:
+            # Make patch array.
+            self.col = conv_nd.im2col_nd_gpu(
+                x, ksize, stride, pad, cover_all=self.cover_all)
+
+            # Compute correlation.
+            W_mat = W.reshape(out_c, -1)
+            col_mats = self.col.reshape(n, -1, reduce(operator.mul, outs))
+            y_mats = y.reshape(n, out_c, -1)
+            # TODO(takagi): Use streams or batch gemm
```

Comment:
CuPy also has [tensordot](https://github.com/pfnet/chainer/blob/ed8ef4786be325899119c3827febc9286da7e354/cupy/linalg/product.py#L129). Can we use it as we use `numpy.tensordot`?